### DISCUSSION TITLE
Tilesets: Workarounds for accessing via a redirect

### DISCUSSION THREAD

**Discussion by TangoYankee** [NYCPlanning/ae-zoning-api#84] [General]
# Description
We ran into an issue accessing the tilesets via a redirect through the API, rather than directly accessing them at their home url. This comes down to a safety feature of browsers. When accessing a url via a redirect, the browser will set the origin to `null`. Because the origin of the request cannot be determined, the CORS policy rejects it and the browser stops the data from being read. 

## First option
The first work-around would be to add `null` to the digital ocean cors policy. This weakens CORS and should be avoided.

## Second option
The second work-around would be to refactor the API endpoint from a redirect to a pass-through. It won't do a simple passthrough; we need to turn off the standard json encoding. This is complicated by the use of pbf files. ~~I haven't been able to find a clear path to encode the data. The [closest I could get](https://github.com/NYCPlanning/ae-zoning-api/blob/4ac75ba0a078eeeee731301e36dbe6008d939180/src/zoning-district/zoning-district.controller.ts#L41) is passing the literal characters, which is not helpful. NestJs does seem to have a [proto library](https://docs.nestjs.com/microservices/grpc). But, I'm not sure how to leverage it for this situation.~~ I applied the code we used for another discussion on protobufs, setting the content-type header to `application/x-protobuf` and sending the response data directly. It seems the `httpService` from nestjs is parsing the protobuf instead of letting it passthrough? In short, I still haven't been able to get it working. example is on [84/exp/tileset-passthrough](https://github.com/NYCPlanning/ae-zoning-api/tree/84/exp/tileset-passthrough)

## Third option
Call the tilesets directly from the frontend

**Comment by TylerMatteo** [comment]
Huh, that is annoying. I'll try to do my own research on this. If we really can't figure out how to do it the way we wanted to, I can't help but feel like option 3 > 2. Actually passing through the data just seems like a lot of burden to put on the API for the sake of semantics/developer experience.

**Comment by TangoYankee** [comment]
**DRAFT**
Please ignore until this notice is removed

### Setup
1. Start **only** the database:  `docker compose up db` or `docker compose start db`
    - the database needs to be configured before martin is built, allowing martin to auto-discover the functions
2. Define the zoning district tileset functions
   - Run the `zoning_district_fill` sql code in `tilesets/zoning-district-fill.sql` through pgadmin/postico/your-preferred-db-manager
   -   - Run the `zoning_district_label` sql code in `tilesets/zoning-district-label.sql` through pgadmin/postico/your-preferred-db-manager
   - **Note** There are several pieces of draft code also in these file. They should not be run. Comments denote the beginning and end of the required code. Run only the code within this block
3. Build and run martin `docker compose build martin` and `docker compose up martin`

**Comment by TangoYankee** [comment]
middleware? https://www.npmjs.com/package/http-proxy-middleware
